ddtrace/bootstrap/sitecustomize.py

```diff
@@ -2,30 +2,110 @@
 Bootstrapping code that is run when using the `ddtrace-run` Python entrypoint
 Add all monkey-patching that needs to run by default here
 """
-import logging
-import os
 import sys
-from typing import Any
-from typing import Dict
 
 
-# Perform gevent patching as early as possible in the application before
-# importing more of the library internals.
-if os.environ.get("DD_GEVENT_PATCH_ALL", "false").lower() in ("true", "1"):
-    import gevent.monkey
+MODULES_LOADED_AT_STARTUP = frozenset(sys.modules.keys())
+MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED = ("gevent",)
 
-    gevent.monkey.patch_all()
 
+import os  # noqa
+
+
+MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging", "typing"}
+if sys.version_info <= (2, 7):
+    MODULES_TO_NOT_CLEANUP |= {"encodings", "codecs"}
+    import imp
+
+    _unloaded_modules = []
+else:
+    import importlib
+
+
+def is_installed(module_name):
+    # https://stackoverflow.com/a/51491863/735204
+    if sys.version_info >= (3, 4):
+        return importlib.util.find_spec(module_name)
+    elif sys.version_info >= (3, 3):
+        return importlib.find_loader(module_name)
+    elif sys.version_info >= (3, 1):
+        return importlib.find_module(module_name)
+    elif sys.version_info >= (2, 7):
+        try:
+            imp.find_module(module_name)
+        except ImportError:
+            return False
+        else:
+            return True
+    return False
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="0").lower()
+    if dd_unload_sitecustomize_modules == "0":
+        return False
+    elif dd_unload_sitecustomize_modules not in ("1", "auto"):
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not any(
+        is_installed(module_name) for module_name in MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED
+    ):
+        return False
+    return True
+
+
+def cleanup_loaded_modules_if_necessary(aggressive=False):
+    if not aggressive and not should_cleanup_loaded_modules():
+        return
+    modules_loaded_since_startup = set(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP)
+    modules_to_cleanup = modules_loaded_since_startup - MODULES_TO_NOT_CLEANUP
```

Comment:
I don't understand why we can't just clean modules up from `set(sys.modules) - MODULES_TO_NOT_CLEANUP`. Or is this to minimize the number of modules we clone (i.e. where `aggressive` comes into play)?

Reply:
@Yun-Kim Figuring out `modules_loaded_since_startup` is necessary because `sys.modules` has more in it than just what's in `import` statements in this file, and unloading some of them can break the interpreter.